site-src/geps/gep-726.md

```diff
@@ -0,0 +1,336 @@
+// Route rule as a HTTPRequestRewrite filter.
+//
+// Support: Extended
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRedirect `json:"path,omitempty"`
+  // ...
+}
+
+// HTTPPathRedirectType defines the type of path redirect.
+type HTTPPathRedirectType string
+
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
+
+  // This type of redirect indicates that any prefix path matches will be
+  // replaced by the substitution value. For example, a path with a prefix match
+  // of "/foo" and a ReplacePrefixPathMatch substitution of "/bar" will have the
+  // "/foo" prefix replaced with "/bar" in matching requests.
+  PathMatchHTTPPathRedirect HTTPPathRedirectType = "ReplacePrefixPathMatch"
+)
+
+// HTTPPathRedirect defines configuration for path redirects.
+type HTTPPathRedirect struct {
+  // Type defines the type of path redirect.
+  //
+  // +kubebuilder:validation:Enum=Absolute;ReplacePrefixPathMatch
+  Type HTTPPathRedirectType `json:"type"`
+
+  // Value defines the HTTP path value to redirect to. An empty value ("")
+  // indicates that the portion of the path to be changed in the redirect should
+  // be removed from the redirect path. For example, a request to "/foo/bar"
+  // with a prefix match of "/foo" would be redirected to "/bar".
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Substitution string `json:"substitution"`
+}
+```
+
+### Rewrites
+
+A new `RequestRewrite` filter can be added to support rewrites. In the following
+example, a a request to `example.com/foo/abc` would be rewritten to
+`example.net/bar/abc`.
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  hostnames:
+  - example.com
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /foo
+      filters:
+      - type: RequestRewrite
+        requestRewrite:
+          hostname: example.net
+          path:
+            type: ReplacePrefixPathMatch
+            substitution: /bar
+```
+
+This would be represent with the following API additions:
+```go
+// HTTPRequestRewrite defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
+//
+// Support: Extended
+type HTTPRequestRewrite struct {
+  // Hostname is the value to be used to replace the Host header value during
+  // forwarding.
+  //
+  // Support: Extended
+  //
+  // +optional
+  // +kubebuilder:validation:MaxLength=255
+  Hostname *string `json:"hostname,omitempty"`
+
+  // Path defines a path rewrite.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRewrite `json:"path,omitempty"`
+}
```

Comment:
There will be two filters: `RequestHeaderModifier` and a `RequestRewrite`, the UX seems a bit off here.
Options are:
- Deprecated RequestHeaderModifier and fold it into this new filter
- Introduce more targetted filters (break this new filter into two): RequestHostnameModifier, RequestPathModifier
- Accept this emergent behavior
- Something else?

Reply:
Although I agree the filters are similar in purpose, I think "Rewrite" is a common and reasonably well understood concept here. In most of the implementations I can find, rewriting is limited to either the request hostname and path or the full URL. There are a small set of implementations that extend "rewrite" terminology to headers, but that does not seem to be nearly as consistent.

I think the term "rewrite" is important here given its widespread use throughout other implementations. I think this is going to be a common term that users are looking for. Although it may technically more consistent to have a `RequestModifier` filter or separate `RequestHostnameModifier` and `RequestPathModifier` filters, I think it would lead to more confusion.

We could also bundle header modification into this proposed rewrite filter, but I'm not sure that provides sufficient value to introduce a breaking change and/or duplicate capabilities.

Reply:
So this is where implementations doing a thing shouldn't be governing our design, right?
The API is to express intent, how the implementations do shouldn't matter - as long as implementations can satisfy the intent.
With that bit, from a purely product and API perspective, would a user not expect a RequestRewrite filter to also be able to modify headers because headers are part of the request?

As a user, wouldn't you be surprised if you had a filter names "RewriteRequest" which won't be able to modify headers? And to modify headers, you had to use a filter called "RequetHeaderModifier"?

Being consistent can help us keep complexity under control in this area.
What are your concerns with duplication of header modification in the RequestRewrite filter as well? We could keep the behavior and intent exactly the same, that should minimize overheads.

Reply:
>The API is to express intent, how the implementations do shouldn't matter - as long as implementations can satisfy the intent.

💯 Agree with this.

>What are your concerns with duplication of header modification in the RequestRewrite filter as well? We could keep the behavior and intent exactly the same, that should minimize overheads.

Any time we provide multiple ways to do something, we have to handle conflicts between those. We also have to take extra effort to document and explain the difference between the ways to do things.

>As a user, wouldn't you be surprised if you had a filter names "RewriteRequest" which won't be able to modify headers? And to modify headers, you had to use a filter called "RequetHeaderModifier"?

It's hard for me to say since I'm already so far into the API design. Some thoughts:

1. Header modification may happen at any point, including before a redirect.
1. The current proposal considers rewriting distinct from redirection, merging these together would lead to more confusion as capabilities diverged.
1. We could consider renaming "RequestRewrite" to "URLRewrite" which would more clearly communicate the proposed scope. That would be somewhat inconsistent with the existing "RequestRedirect" filter name, but this is also smaller in scope.
1. We could consider moving everything into smaller filters - PathModifier, HostnameModifier, HeaderModifier, SchemeModifier, etc and then have one final filter that indicated all of the above should result in a redirect and not just a rewrite. I'm concerned that this could result in unnecessary confusion around conformance levels + breaking API changes.

Reply:
Chatted with @hbagdi a bit and decided that 3 was the most reasonable approach going forward, I've updated the GEP to reflect that. I've also created https://github.com/kubernetes-sigs/gateway-api/issues/928 as a related follow up from our conversation.